types/decimal.go

```diff
@@ -0,0 +1,324 @@
+// nil embedded big.Int
+type Dec struct {
+	*big.Int `json:"int"`
+}
+
+// number of decimal places
+const Precision = 10
+
+func precisionInt() *big.Int {
+	return new(big.Int).Exp(big.NewInt(10), big.NewInt(Precision), nil)
+}
+
+// nolint - common values
+func ZeroDec() Dec { return Dec{big.NewInt(0)} }
+func OneDec() Dec  { return Dec{precisionInt()} }
+
+// get the precision multiplier
+func precisionMultiplier(prec int64) *big.Int {
+	if prec > Precision {
+		panic("too much precision")
+	}
+	zerosToAdd := Precision - prec
+	multiplier := new(big.Int).Exp(big.NewInt(10), big.NewInt(zerosToAdd), nil)
+	return multiplier
+}
+
+// create a new Dec from integer assuming whole numbers
+// CONTRACT: prec !> Precision
+func NewDec(i, prec int64) Dec {
+	return Dec{
+		new(big.Int).Mul(big.NewInt(i), precisionMultiplier(prec)),
+	}
+}
+
+// create a new Dec from big integer assuming whole numbers
+// CONTRACT: prec !> Precision
+func NewDecFromBigInt(i *big.Int, prec int64) Dec {
+	return Dec{
+		new(big.Int).Mul(i, precisionMultiplier(prec)),
+	}
+}
+
+// create a new Dec from big integer assuming whole numbers
+// CONTRACT: prec !> Precision
+func NewDecFromInt(i Int, prec int64) Dec {
+	return Dec{
+		new(big.Int).Mul(i.BigInt(), precisionMultiplier(prec)),
+	}
+}
+
+// create a decimal from a decimal string (ex. "1234.5678")
+func NewDecFromStr(str string) (d Dec, err Error) {
+	if len(str) == 0 {
+		return d, ErrUnknownRequest("decimal string is empty")
+	}
+
+	// first extract any negative symbol
+	neg := false
+	if string(str[0]) == "-" {
+		neg = true
+		str = str[1:]
+	}
+
+	if len(str) == 0 {
+		return d, ErrUnknownRequest("decimal string is empty")
+	}
+
+	strs := strings.Split(str, ".")
+	lenDecs := 0
+	combinedStr := strs[0]
+	if len(strs) == 2 {
+		lenDecs = len(strs[1])
+		if lenDecs == 0 || len(combinedStr) == 0 {
+			return d, ErrUnknownRequest("bad decimal length")
+		}
+		combinedStr = combinedStr + strs[1]
+	} else if len(strs) > 2 {
+		return d, ErrUnknownRequest("too many periods to be a decimal string")
+	}
+
+	if lenDecs > Precision {
+		return d, ErrUnknownRequest("too much Precision in decimal")
+	}
+
+	// add some extra zero's to correct to the Precision factor
+	zerosToAdd := Precision - lenDecs
+	zeros := fmt.Sprintf(`%0`+strconv.Itoa(zerosToAdd)+`s`, "")
+	combinedStr = combinedStr + zeros
+
+	combined, ok := new(big.Int).SetString(combinedStr, 10)
+	if !ok {
+		return d, ErrUnknownRequest("bad string to integer conversion")
+	}
+	if neg {
+		combined = new(big.Int).Neg(combined)
+	}
+	return Dec{combined}, nil
+}
+
+//nolint
+func (d Dec) IsZero() bool      { return (d.Int).Sign() == 0 } // Is equal to zero
+func (d Dec) Equal(d2 Dec) bool { return (d.Int).Cmp(d2.Int) == 0 }
+func (d Dec) GT(d2 Dec) bool    { return (d.Int).Cmp(d2.Int) == 1 }             // greater than
+func (d Dec) GTE(d2 Dec) bool   { return !d.LT(d2) }                            // greater than or equal
```

Comment:
Can this be switched to 
`(d.Int).Cmp(d2.Int) > -1`?
I'd prefer to not use a function call here, for better efficiency as its reasonable to assume that decimals will be in tight loops.

Reply:
oh yeah this is a good idea.. thanks! included